### ISSUE TITLE
[T-SQL] AM07 False Positive: Parser conflates INSERT and UNION SELECT when semicolon is missing

### ISSUE THREAD

**Issue by Zamiell** [sqlfluff/sqlfluff#7326]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I encountered a false positive AM07 (ambiguous.set_columns) error when linting T-SQL.

The parser appears to fail to terminate an INSERT INTO ... SELECT statement when it is followed by a UNION ALL statement if the INSERT is not explicitly ended with a semicolon. It seems to treat the SELECT inside the INSERT (which has 4 columns in the example below) as the first query in the UNION set, comparing it against the subsequent SELECT (which has 3 columns), resulting in a column count mismatch.

Adding a semicolon after the INSERT statement resolves the issue, but the linter should ideally parse the statement boundaries correctly or raise a parsing error rather than an ambiguity rule violation.

Steps to Reproduce:

Save the following content to reprex.sql:

```sql
CREATE PROCEDURE [dbo].[usp_ReprexParsingBug]
    @LstDiagnosis [dbo].[DiagnosisList] READONLY
AS
BEGIN
    CREATE TABLE #DiagRestrictionIDs(
        ID INT IDENTITY(1,1),
        DiagnosisRestrictionID BIGINT,
        DiagnosisDistinction VARCHAR(20),
        IsRestrictionType BIT,
        ConfigureMessage varchar(8000)
    )

    -- 1. INSERT statement with 4 columns
    INSERT INTO #DiagRestrictionIDs(DiagnosisRestrictionID, DiagnosisDistinction,IsRestrictionType,ConfigureMessage)
    SELECT
        mdr.ID,
        mdr.DiagnosisDistinction,
        mdr.IsRestrictionType,
        mdr.ConfigureMessage
    FROM UnknownTable_Source mdrc
    INNER JOIN UnknownTable_Restrictions mdr ON mdr.ID = mdrc.ID
    GROUP BY mdr.ID, mdr.DiagnosisDistinction, mdr.IsRestrictionType, mdr.ConfigureMessage
    -- MISSING SEMICOLON HERE causes the parser to merge this with the next query

    -- 2. UNION query with 3 columns (Triggering AM07)
    SELECT
        lst.Diagnosis,
        mdr.IsRestrictionType,
        mdr.ConfigureMessage
    FROM #DiagRestrictionIDs mdr
    INNER JOIN UnknownTable_Codes drd ON mdr.DiagnosisRestrictionID = drd.MdDiagnosisCongifurationId
    INNER JOIN @LstDiagnosis lst ON drd.MdDiagnosisCodeId = lst.MDDiagnosisID

    UNION ALL

    SELECT
        lst.Diagnosis,
        mdr.IsRestrictionType,
        mdr.ConfigureMessage
    FROM #DiagRestrictionIDs mdr
    LEFT JOIN @LstDiagnosis lst ON 1=1
END
```

Then lint it with: `sqlfluff lint --dialect tsql --rules AM07 reprex.sql`

```
== [reprex.sql] FAIL
L:  15 | P:  5 | AM07 | Queries within set query produce different numbers of
                       | columns. [ambiguous.set_columns]
```

Expected Behaviour: The linter should recognize the end of the INSERT statement and treat the subsequent SELECT as a new statement. It should pass without AM07 errors (or raise a different error regarding missing semicolons if that rule is enabled).

Observed Behaviour: The linter raises AM07, incorrectly identifying a column count mismatch between the INSERT statement and the UNION query.

Versions:
python: 3.13.3

sqlfluff: 3.4.2

dialect: tsql

Relevant Configuration: n/a

Finally, note that this bug can be worked around by adding a semicolon, but this is still a bug in the parser and/or sqlfluff.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by Zamiell** [comment]
I asked Claude to take a look, and he reported:


##  Attempted Fixes

  I've tried several approaches:

  1. Using NonSetSelectableGrammar in INSERT - This prevents INSERT from matching bare set expressions, but it breaks parsing of stored procedures that have both INSERT and UNION statements
  2. Adding terminators to UnorderedSelectStatementSegment - This doesn't prevent SetExpressionSegment from continuing to match after the terminator
  3. Various terminator configurations - None worked correctly

##  The Core Challenge

  T-SQL doesn't require semicolons to separate statements. The parser relies on grammar matching to determine statement boundaries, but in this case:
  - SetExpressionSegment matches: SELECT ... (from INSERT) + UNION ALL + SELECT ... (next statement)
  - The parser can't distinguish between:
    - INSERT INTO t SELECT ... UNION SELECT ... (one statement with UNION in INSERT)
    - INSERT INTO t SELECT ... + SELECT ... UNION SELECT ... (two separate statements)

**Comment by Zamiell** [comment]
I've tried many approaches and they all break the parser. The issue is that this is a very difficult problem in T-SQL due to optional semicolons.

  1. Created InsertSelectableGrammar - Failed, broke BEGIN...END parsing
  2. Used NonSetSelectableGrammar directly in INSERT - Failed, broke BEGIN...END parsing
  3. Used SelectStatementSegment in INSERT - Failed, broke BEGIN...END parsing
  4. Reordered NonWithSelectableGrammar - Failed, still greedily consumed UNION
  5. Modified SelectableGrammar globally - Failed, broke BEGIN...END parsing

  The fundamental challenge: T-SQL doesn't require semicolons, so the parser uses greedy matching to determine statement boundaries. Any restriction I add to prevent INSERT from consuming following UNION statements breaks the parser's ability to handle the test case.

  The root cause appears to be that SetExpressionSegment uses AnyNumberOf with greedy matching, and there's no clean way to add terminators in T-SQL since semicolons are optional.

Is there someone here with deep T-SQL parser knowledge who can suggest a solution?

**Comment by peterbud** [comment]
I'll look into that

**Comment by peterbud** [comment]
Just looked at it briefly:
* I can confirm the above query produces the issue
* Interestingly if you remove / comment out the GROUP BY line, it properly parses/lints - @Zamiell can you check? I'm not sure yet, but that hints that the issue might be with the GROUP BY clause syntax in the parser.
* I need to reduce the example to a minimalist example which shows the issue

**Comment by peterbud** [comment]
OK, here is the minimal repro query:
```sql
-- QUERY 1
SELECT FIELD2
FROM TABLE2
GROUP BY
    FIELD3

-- MISSING SEMICOLON HERE causes 
-- the parser to merge this with the next query

-- QUERY 2. UNION query with 2 columns
SELECT
    FIELD1,
    FIELD2
FROM TABLE2

UNION ALL

SELECT
    FIELD3,
    FIELD4
FROM TABLE3
```

I will submit tomorrow the fix, it is the `GroupByClauseSegment`, which is faulty
